File: JavaServer/src/Server/StringModificationTree.java
package Server;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

public class StringModificationTree {
	
	public static void main( String[] args ){
		StringModificationTree string = new StringModificationTree( System.currentTimeMillis()-1000, "" );
		
		string.applyTransformation( new StringNode( System.currentTimeMillis(), 0, "Hello World" ) );
		string.applyTransformation( new StringNode( System.currentTimeMillis()-500, 0, "a" ) );
		
		
		StringNode curr_node = string.root_node;
		// Find the last single modification moment in the tree
		while( curr_node.children.size() > 0 ){
			curr_node = curr_node.children.get(0);
		}
		
		string.simplify();
		
		System.out.println( string.root_node.data );
	}
	
	public static final int ADDITION = 1;
	public static final int DELETION = 2;
	
	StringNode root_node;
	
	public StringModificationTree( long start_time, String base_string ){
		root_node = new StringNode( start_time, 0, base_string );
	}
	
	public String toString(){
		simplify();
		return root_node.data;
	}
	
	public void applyTransformation( StringNode node ){
		StringNode curr_node = root_node;
		
		// Find the last single modification moment in the tree
		while( curr_node.children.size() > 0 ){
			if( curr_node.children.size() == 1 ){
				if( curr_node.children.get(0).mod_time <= node.mod_time ){
					curr_node = curr_node.children.get(0);
				}else{
					node.children.add(curr_node.children.get(0));
					curr_node.children.clear();
					break;
				}
			}else{
				break;
			}
		}
		
		curr_node.children.add(node);
	}
	
	private void simplify(){
		// While the root node has children
		while( !root_node.children.isEmpty() ){
			// If there is only one child
			switch( root_node.children.size() ){
			case( 1 ): // Add the one child node to the parent node
				root_node = update( root_node, root_node.children.get(0) );	
				break;
			
			default: // merge the two or more children nodes
				root_node = merge( root_node, root_node.children );
				break;
			}
		}
	}
	
	private StringNode merge(StringNode base_node, List<StringNode> children) {
		Collections.sort(children);
		StringNode result = base_node;
		
		for( int i=0 ; i<children.size() ; i++ ){
			// Apply this transform
			result = update( result, children.get(i) );
			
			// Adjust all other transforms relative
			for( int j=i ; j<children.size() ; j++ ){
				children.get(j).applyTransform( children.get(i) );
			}
		}
		
		return result;
	}

	private StringNode update(StringNode base_node, StringNode mod_node) {
		String result_string = base_node.data;
		
		switch( mod_node.mod_type ){
		case( ADDITION ):
			result_string = result_string.substring( 0, mod_node.mod_loc ) + mod_node.data + result_string.substring( mod_node.mod_loc );
			break;
		
		case( DELETION ):
			result_string = result_string.substring( 0, mod_node.mod_loc ) + result_string.substring( mod_node.mod_loc + mod_node.length );
			break;
		}
		
		StringNode new_node = new StringNode( mod_node.mod_time, 0, result_string );
		new_node.children = mod_node.children;
		
		return new_node;
	}

	public static class StringNode implements Comparable<StringNode> {
		
		public int mod_type = 0;
		public long mod_time;
		public int mod_loc;
		public int length = 0;
        public String data;
        public StringNode parent = null;
        public List< StringNode > children = new LinkedList<>();
		
		public StringNode( long mod_time, int mod_val, String new_val ) {
			this.mod_time = mod_time;
			data = new_val;
			mod_loc = mod_val;
			
			mod_type = ADDITION;
		}
		
		/**
		 * Apply transform node before this transform
		 * @param node
		 */
		public void applyTransform(StringNode node) {
			int a = mod_loc;
			int b = a + length;
			int c = node.mod_loc;
			int d = c + node.mod_loc;
			
			if( b < c ){ // to the left of node
				// do nothing
				
			}else if( b < d ){
				if( a > c ){ // inside of node
					mod_loc += node.length;
					
				}else{ // left overlap with node
					// do nothing
					
				}
			}else if( a < d ){ // right overlap with node
				mod_loc += node.length;
				
			}else{ // to the right of node
				mod_loc += node.length;
				
			}
		}

		public StringNode( long mod_time, int mod_val, int length ) {
			this.mod_time = mod_time;
			this.length = length;
			
			mod_type = DELETION;
		}

		@Override
		public int compareTo(StringNode o1) {
			long o1_t = o1.mod_time;
			
			if( o1_t == mod_time ){
				return 0;
			}else if( mod_time < o1_t ){
				return -1;
			}else{
				return 1;
			}
		}
		
	}
	
}

File: PythonPlugin/StringModificationTree.py
ADDITION = 1
DELETION = 2

class StringModificationTree(object):
	def __init__(self, start_time, base_string):
		self.root_node = StringNode(start_time, 0, base_string)

	def toString(self):
		simplify()
		return self.root_node.data

	def applyTransformation(self, node):
		curr_node = self.root_node

		#Find the last single modification moment in the tree
		while len(curr_node.children) > 0:
			if len(curr_node.children) == 1:
				if curr_node.children[0].mod_time <= node.mod_time:
					curr_node = curr_node.children[0]
				else:
					node.children.append(curr_node.children[0])
					del curr_node.children[:]
					break
			else:
				break

		curr_node.children.append(node)

	def simplify(self):
		while len(self.root_node.children) != 0:
			if len(self.root_node.children) == 1:
				self.root_node = update(self.root_node, self.root_node.children[0])
				break
			else:
				self.root_node = merge(self.root_node, self.root_node.children)
				break

	def merge(self, base_node, children):
		result = base_node

		for i in range(0, len(children)):
			result = update(result, children[i])

			for j in range(i, len(children)):
				children[j].applyTransform(children[i])

		return result

	def update(self, base_node, mod_node):
		result_string = base_node.data

		if mod_node.mod_type == ADDITION:
			result_string = result_string[0:mod_node.mod_loc] 
			+ mod_node.data + result_string[mod_node.mod_loc:]
			
		elif mod_node.mod_type == DELETION:
			result_string = result_string[0:mod_node.mod_loc]
			+ result_string[mod_node.mod_loc + mod_node.length:]
			
		
		new_node = StringNode(mod_node.mod_time, 0, result_string)
		new_node.children = mod_node.children

		return new_node

class StringNode(object):
	mod_type = 0
	length = 0
	children = []

	def __init__(self, mod_time, mod_val, new_val):
		self.mod_time = mod_time
		self.data = new_val
		self.mod_loc = mod_val

		self.mod_type = ADDITION

	def __init__(self, mod_time, mod_val, length, haha):
		self.mod_time = mod_time
		self.length = length

		self.mod_type = DELETION

	def compareTo(self, o1):
		self.o1_t = o1.mod_time

		if o1_t == self.mod_time:
			return 0
		elif self.mod_time < o1_t:
			return -1
		else:
			return 1

	def applyTransform(self, node):
		a = node.mod_loc
		b = a + node.length
		c = self.mod_loc
		d = c + self.mod_loc
		if node.mod_type == ADDITION:					# first operation is ADDITION	
			if self.mod_type == ADDITION:					# second operation is ADDITION done
				if b <= c:
					self.mod_loc += node.length
			else											# second operation is DELETION done
				if b <= c:
					self.mod_loc -= node.length
				elif a >= c && a < d:
					self.length += node.length
		else:											# first operation is DELETION
			if self.mod_type == ADDITION:					# second operation is ADDITION 
				if b <= c:
					self.mod_loc -= node.length
				elif a < c:
					self.mod_loc -= c - a
			else											# second operation is DELETION done
				if b <= c:
					self.mod_loc -= node.length
				elif a <= c:
					if b < d:
						self.mod_loc -= c - a
						self.length -= b - c
					else:
						self.length = 0
				elif a < d:
					if b < d:
						self.length -= b - a
					else:
						self.length -= d - a


